server/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -1992,70 +1989,54 @@ protected final Releasable acquireEnsureOpenRef() {
 
     /**
      * When called for the first time, puts the engine into a closing state in which further calls to {@link #acquireEnsureOpenRef()} will
-     * fail with an {@link AlreadyClosedException} and waits for all outstanding ensure-open refs to be released, before returning {@code
-     * true}. If called again, returns {@code false} without waiting.
+     * fail with an {@link AlreadyClosedException} and waits for all outstanding ensure-open refs to be released. If called again, returns
+     * without waiting.
      *
-     * @return a flag indicating whether this was the first call or not.
+     * @param listener the action to perform when all open refs are released
      */
-    private boolean drainForClose() {
+    private void drainForClose(ActionListener<Void> listener) {
         if (isClosing.compareAndSet(false, true) == false) {
             logger.trace("drainForClose(): already closing");
-            return false;
+            return;
         }
 
         logger.debug("drainForClose(): draining ops");
         releaseEnsureOpenRef.close();
-        final var future = new UnsafePlainActionFuture<Void>(ThreadPool.Names.GENERIC) {
-            @Override
-            protected boolean blockingAllowed() {
-                // TODO remove this blocking, or at least do it elsewhere, see https://github.com/elastic/elasticsearch/issues/89821
-                return Thread.currentThread().getName().contains(ClusterApplierService.CLUSTER_UPDATE_THREAD_NAME)
-                    || super.blockingAllowed();
-            }
-        };
-        drainOnCloseListener.addListener(future);
-        try {
-            future.get();
-            return true;
-        } catch (ExecutionException e) {
-            logger.error("failure while draining operations on close", e);
-            assert false : e;
-            throw new IllegalStateException(e);
-        } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-            logger.error("interrupted while draining operations on close");
-            throw new IllegalStateException(e);
-        }
+        drainOnCloseListener.addListener(listener);
     }
 
     /**
      * Flush the engine (committing segments to disk and truncating the translog) and close it.
      */
     public void flushAndClose() throws IOException {
```

Comment:
Are you sure it works to leave this method as synchronous? It's called from `org.elasticsearch.index.shard.IndexShard#close` and when it returns it will execute the subsequent `finally` block, which may now be happening too soon while it's still draining.

Reply:
Yes, that's a very good point. It seems there's indeed an expectation that `Engine#flushAndClose` and `Engine#close` are synchronous. I wonder if we need to make them asynchronous too, because otherwise we would still need to use `PlainActionFuture`